### DISCUSSION TITLE
Elixir driver for EdgeDB

### DISCUSSION THREAD

**Discussion by nsidnev** [geldata/gel#2549] [Show and tell]
Hi!

I made a binary driver for the Elixir language. It's not fully ready right now, but its core, it seems to me, is done.
It can connect to the EdgeDB instance, prepare and execute queries and parse results from database. It also supports encoding of Elixir types into EdgeDB types and vice verce. Right now it lacks support of transactions, has no tests, the code needs some cleanup, and public API isn't defined, but it works :)

I'll be glad to hear any feedback from EdgeDB team :)

Link to the repository: https://github.com/nsidnev/edgedb-elixir

**Comment by colinhacks** [comment]
Wow this is some phenomenal work! Might take us some time to look through the codebase since we don't use much Elixir (yet). In the meantime we'll do our best to point other Elixir users towards this project. Great stuff! 👍

**Reply by nsidnev** [reply]
> Wow this is some phenomenal work!

Thanks! :)

Over the past few days, I've improved the driver a bit:
1. added raw transactions (retrying transactions are in progress)! 🎉
2. added basic tests for all (?) codecs and authentication
3. added new public API which tries to follow API of existing drivers:
```elixir
{:ok, conn} = EdgeDB.start_link(port: 10_700)
42 = EdgeDB.query_one!(conn, "SELECT <int64>42")
```
4. added execution of queries using an `optimistic execute` flow
5. configured CI to run tests on EdgeDB
6. several other fixes and improvements

---

BTW. Is there a way to setup CI using `edgedb/edgedb-setup` without manual running `edgedb-server` and other tools?
I've looked at the `edgedb-go` sources and it does all bootstrapping stuff manually in test script. IMO this is not very convenient. 

In my opinion the ideal way would be something like this:
1. use GitHub Actions' `services` to run EdgeDB as a docker container
2. use `edgedb/edgedb-setup` action to install CLI
3. somehow (I'm not sure if this is possible right now) use `edgedb projects` to link a running EdgeDB instance with one from `edgedb.toml` config

**Reply by elprans** [reply]
> BTW. Is there a way to setup CI using `edgedb/edgedb-setup` without manual running `edgedb-server` and other tools?

Not right now, but there's a plan to improve the action and enable integration with `edgedb project`

**Reply by nsidnev** [reply]
Awesome! Can't wait for this :)

**Comment by nsidnev** [comment]
Hi! This topic was abandoned for a while, but I'd like to bring it up again :)

All this time I've been following the development of official drivers for EdgeDB and tried to add new features of your drivers to the Elixir driver. And right now most of features are available there, thanks to your implementations ❤️. 

I plan to document and release it in the Elixir package manager when EdgeDB is officially released! 🚀

There is also a small project that shows how to use the driver until the documentation is added. It can be deployed to `fly.io` using your guide and a few additional commands for the app itself :)
Link to the example project: https://github.com/nsidnev/edgedb-phoenix-example

**Reply by 1st1** [reply]
Wow!!! This is great, Nik! 🚀
